format/spec.md

```diff
@@ -1605,13 +1611,8 @@ All readers are required to read tables with unknown partition transforms, ignor
 Writing v3 metadata:
 
 * Partition Field and Sort Field JSON:
-    * `source-ids` was added and is required
-    * `source-id` is no longer required and should be omitted; always use `source-ids` instead
-
-Reading v1 or v2 metadata for v3:
-
-* Partition Field and Sort Field JSON:
-    * `source-ids` should default to a single-value list of the value of `source-id`
+    * `source-ids` was added and is required in the case of multi-argument transforms.
+    * `source-id` should still be written in the case of single-argument transforms.
```

Comment:
This sounds reasonable to me, except that I think `source-ids` should be required all the time.

Reply:
Why not only writing `source-ids`, even for single argument (single value list) ?

Reply:
Just to clarify, do you mean also producing `source-ids` when `TableVersion≤2`?

Reply:
No, I mean that TableVersion 3 should create both `source-id` and `source-ids` (for backward compatibility).

My initial idea was to only use `source-ids` for TableVersion 3 (even for single arg transform), but it would not work with reader from previous version.

Reply:
Iceberg only supports forward compatibility, so V3 readers should be able to read V2, but not the other way around. Still, requiring the table version for serializing the partition-spec does not sound ideal to me. I've changed the PR back to depend on the serialization based on the number of arguments, rather than the table version.

Reply:
I thought from the other discussion we are now changing this to source-id is required for single arg transforms and source-ids is required for multi arg. So both are optional but are still "must write" language in the spec

Reply:
I agree with the current changes to use both `source-id` and `source-ids` depending on whether the transform is multi-arg.

Reply:
I agree, I just think this particular line makes it sound optional so I wanted to remove the word `should`

Reply:
@RussellSpitzer I agree, thanks for pointing out. I've updated the wording, let me know what you think!